README.md
# GMail::Utils

GmVault is a small utility to backup all or parts of a given GMail mailbox.  It generates
a set of directories organized by date (YYYY-MM/*).  Each folder has two files per mail, one
.eml containing the actual mail (headers and body) and a .meta containing all GMail metadata
(id of the mail, date sent, subject, thread ID for a given conversation and so on).

This gem is composed of a library (GMail::Utils) and the corresponding script to transform that
backup into a proper Maildir-formatted mailbox.  The script takes a tag as main argument and
will take only mails matching the given tag.

## Installation

Add this line to your application's Gemfile:

    gem 'gmail-utils'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install gmail-utils

## Usage

    tag2maildir [-h] [-b DIR] -t TAG DIR

    -t TAG      Use the following tag +TAG+ to filter the mails
    -b DIR      Use the given directory as base dir to create the mailbox (defaults to +~/Maildir+)
    DIR         Use the following path to the gmvault directory

    $Id: README.md,v af7380ec209e 2012/09/24 22:14:17 roberto $


ChangeLog
2012-12-02  Ollivier Robert  <roberto@keltia.net>

	* spec/config_spec.rb:
	More specs for GMail::Config.
	[a133689cd887] [tip]

	* lib/gmail/config.rb:
	More yard conments.
	[034e1e09cbd0]

	* test/Perso-Foo/index.tch:
	Sync.
	[73ef7f4dda49]

	* lib/gmail/config.rb:
	Move BASE_DIR definition within the class.
	[ab6b367be5cf]

	* lib/gmail/index.rb:
	Classic = vs == blunder. Fix.
	[2c42bdbd49e3]

2012-11-25  Ollivier Robert  <roberto@keltia.net>

	* test/Perso-Foo/index.tch:
	Update test db.
	[348014b128fb]

	* spec/index_spec.rb:
	Add specs for different db.
	[d28f37d11fce]

	* lib/gmail/index.rb:
	Fix last_sync usage.
	[e5d6e5a6b22f]

	* lib/gmail/version.rb, test/Perso-Foo/index.tch:
	Update to 0.4.7 & DB v3.
	[afb8ea71b79f]

	* test/Perso-Foo/index.tch:
	Update test db.
	[9ce96edbfc0e]

	* bin/tag2maildir:
	Update tag2maildir to DB v3.
	[abb9c5b2fd7b]

	* lib/gmail/index.rb:
	Upgrade DB to 3.
	[dc5af50c1862]

	* test/Perso-Foo/index.tch:
	Update local test index.
	[5d5832ea5531]

	* lib/gmail/index.rb, spec/index_spec.rb:
	Add #last_sync & #last_sync= + specs.
	[11c93bfca476]

2012-11-08  Ollivier Robert  <roberto@keltia.net>

	* Automated merge with ssh://centre.keltia.net//data/HG/mirrors/gmail-
	utils
	[b43ed56a509a]

	* share/gmail-backup.sh:
	Merge TODO entry.
	[592034610789]

2012-11-06  Ollivier Robert  <roberto@keltia.net>

	* Automated merge with ssh://centre.keltia.net//data/HG/mirrors/gmail-
	utils
	[303040ebb22a]

2012-10-31  Ollivier Robert  <roberto@keltia.net>

	* share/gmail-backup.sh:
	Mention a TODO item.
	[a7821bf8be6d]

2012-11-07  Ollivier Robert  <roberto@keltia.net>

	* Automated merge with ssh://centre.keltia.net//data/HG/mirrors/gmail-
	utils
	[d5f54933b449]

	* Automated merge with ssh://centre.keltia.net//data/HG/mirrors/gmail-
	utils
	[9b57c31407da]

	* share/gmail-backup.sh:
	Try to fix tsocks support.
	[f23344581031]

2012-11-03  Ollivier Robert  <roberto@keltia.net>

	* spec/index_spec.rb:
	Move nil test before the others.
	[6de8415ee19d]

2012-11-01  Ollivier Robert  <roberto@keltia.net>

	* spec/index_spec.rb:
	Remove unneeded setting.
	[f4193721a614]

	* bin/tag2maildir:
	Use GMail::Index#version.
	[20ea3b7deb76]

	* lib/gmail/index.rb, spec/index_spec.rb:
	Add GMail::Index#version + specs.
	[9e9f0ea1ecf3]

	* bin/tag2maildir:
	Reformat some lines, displaying gmvault directory.
	[10e9be32a427]

	* lib/gmail/tag.rb:
	return is useless here.
	[322d1be5001a]

	* lib/gmail/entity.rb:
	then is useless here.
	[0ed856a82bc2]

	* bin/tag2maildir:
	Skip invalid mails.
	[8a45925e1fc1]

2012-10-31  Ollivier Robert  <roberto@keltia.net>

	* bin/tag2maildir:
	Merge the move of initialization outside loop.
	[34eeed2f7dad]

	* doc/GMail.html, doc/GMail/Config.html, doc/GMail/Entity.html,
	doc/GMail/Index.html, doc/GMail/Tag.html, doc/GMail/Utils.html,
	doc/_index.html, doc/class_list.html, doc/file.README.html,
	doc/frames.html, doc/index.html, doc/method_list.html, doc/top-
	level-namespace.html:
	Update documentation.
	[6ea4d431a145]

	* .hgtags:
	Added tag v_0_4_6 for changeset 2fb5fbe8fd00
	[8d017eaf449d]

	* lib/gmail/version.rb:
	Bump version to 0.4.6.
	[2fb5fbe8fd00] [v_0_4_6]

	* bin/tag2maildir, gmail-utils.gemspec, lib/gmail-utils.rb, lib/gmail-
	utils/version.rb, lib/gmail.rb, lib/gmail/index.rb,
	lib/gmail/version.rb:
	Move gmail-utils/version.rb into gmail/.
	[64d724f6c964]

	* lib/gmail-utils/version.rb, lib/gmail/index.rb:
	Add a version number to the Index database.
	[a37d0dd5a275]

	* lib/gmail/index.rb:
	Fix comment.
	[029b499b1b5b]

	* test/Perso-Foo/index.tch:
	Update the db version.
	[b243d38967b2]

2012-10-27  Ollivier Robert  <roberto@keltia.net>

	* bin/tag2maildir:
	Use the gmdb path index stored in the index (save it if not).
	Reorder initialization.
	[d911d2829b92]

	* .hgtags:
	Added tag v_0_4_5 for changeset adfa9463b401
	[52fb062d963b]

	* lib/gmail-utils/version.rb:
	Bump version to 0.4.5.
	[adfa9463b401] [v_0_4_5]

	* test/Perso-Foo/index.tch:
	gmdb is now saved in index.tch.
	[de5af34d35bb]

	* bin/tag2maildir:
	Use the newly implemented Index#gmdb=
	[24b716f83b96]

	* lib/gmail/index.rb, spec/index_spec.rb:
	Implement specs and code for #gmdb & #gmdb= in Index class.
	[49d0c765ad5c]

2012-10-23  Ollivier Robert  <roberto@keltia.net>

	* lib/gmail/config.rb, spec/config_spec.rb,
	test/dot.gmvault/keltia+fake@gmail.com.oauth:
	Specs, code and test data for GMail::Config.
	[341f2ce15d43]

2012-10-22  Ollivier Robert  <roberto@keltia.net>

	* lib/gmail/config.rb, spec/config_spec.rb:
	Begin to implement GMail::Config to read gmvault config.
	[d710b2d69339]

	* lib/gmail.rb:
	Add loading of GMail::Config.
	[e6575f415994]

	* bin/tag2maildir:
	Replace displaying the $Id revision by the current tag.
	[cab4d9edc46c]

2012-10-21  Ollivier Robert  <roberto@keltia.net>

	* lib/gmail-utils/version.rb:
	Utils is a module, not a class.
	[82f8e31a126b]

2012-10-19  Ollivier Robert  <roberto@keltia.net>

	* .hgtags:
	Added tag v_0_4_4 for changeset e325783196a5
	[f5db6647c2f1]

	* bin/gmail-backup, lib/gmail-utils/version.rb, share/gmail-backup.sh:
	Fix the way we find the shell script, bump version.
	[e325783196a5] [v_0_4_4]

2012-10-18  Ollivier Robert  <roberto@keltia.net>

	* bin/gmail-backup, lib/gmail/core.rb:
	Fix module name.
	[5a113e62ab59]

	* bin/gmail-backup:
	Remove spurious parenthesis.
	[ce01211c39c8]

	* bin/gmail-backup, lib/gmail.rb, lib/gmail/core.rb:
	Use the same method as Raw::Utils to know where we are.
	[004b15a184a6]

	* .hgtags:
	Added tag v_0_4_3 for changeset d226231e0971
	[ac7c46e01a8d]

	* lib/gmail-utils/version.rb:
	Bump it to 0.4.3 with addition of gmail-backup.
	[d226231e0971] [v_0_4_3]

	* share/gmail-backup.sh:
	Add the shell script.
	[38db2fb81eac]

	* bin/gmail-backup:
	Add a Ruby wrapper for the shell script.
	[7687718141dc]

	* bin/gmail-backup:
	A Shell script is not correctly seen as such.
	[cd06e57b99dc]

	* .hgignore:
	Add .bundle to the ignore list.
	[820df62ba1a7]

	* bin/gmail-backup:

lib/gmail/index.rb
# @abstract Manages an index of message-id in a given Maildir mailbox
#
# @author Ollivier Robert <roberto@keltia.net>

VCS_GMI_ID = '$Id: index.rb,v c4e297bd63db 2012/12/06 15:07:09 roberto $'

require 'rufus/tokyo'
require 'gmail/version'

# Main namespace
module GMail
  # Manages a database of message-ids from mail converted from gmvault into Maildir
    class Index
    attr_reader :path, :db

    # Create a new index if the file is not present or open the current one.
    # @param [String] path path to the mailbox in Maildir format
    def initialize(path)
      if File.exists?("#{path}/cur") and File.exists?("#{path}/new")
        @path = path
      else
        raise ArgumentError, "#{path} is not a Maildir mailbox"
      end
      @db = Rufus::Tokyo::Cabinet.new("#{path}/index.tch")
      check_db()
    end

    # Check db version and upgrade it if needed
    def check_db
      if @db['db_version'].nil?
          @db['db_version'] = GMail::Utils::DB_VERSION
      end
      if @db['gmdb'].nil?
        puts('Warning: gmdb not found...')
      end
      if @db['db_version'] == '2'
        @db['last_sync'] = Time.now
        @db['db_version'] = '3'
      end
    end

    # Check if a given gm_id is present
    # @param gm_id GMail ID of the message
    # @return [TrueClass|FalseClass] true if present
    def present?(gm_id)
      return false if gm_id.nil?
      return false if @db[gm_id].nil?
      true
    end

    # Allow getting a value
    # @param gm_id GMail ID of the message
    # @return [TrueClass|FalseClass] whether it has been stored
    def [](gm_id)
      if gm_id.nil?
        return(gm_id)
      end
      @db[gm_id]
    end

    # Allow inserting a new value
    # @param gm_id GMail ID of the message
    # @param value Value to assign
    def []=(gm_id, value)
      return nil if gm_id.nil?
      @db[gm_id] = value
    end

    # Return the last_id from the db
    # @return [String] +last_id'+property from the db
    def last_id
      @db['last_id']
    end

    # Set the last_id property in the db
    # @param [Fixnum] value set the +last_id+ property to +value+
    def last_id=(value)
      @db['last_id'] = value
    end

    # Return the time of the last sync through the 'last_sync' property
    # @return [Time] +last_sync+ property from the db
    def last_sync
      @db['last_sync']
    end

    # Set the last_sync property in the db
    # @param [Fixnum] value set the +last_id+ property to +value+
    def last_sync=(value)
      @db['last_sync'] = value
    end

    # Returns the size of the db
    # @return [Fixnum] size of the db
    def size
      @db.size
    end

    # Returns the path to the gmvault db
    # @return [String] path to gmvault db
    def gmdb
      @db['gmdb']
    end

    # Set the gmvault db path in the index
    # @param [String] value path to the gmvault db
    def gmdb=(value)
      @db['gmdb'] = value || ''
    end

    # Returns the db version
    # @return [Fixnum] version number
    def version
      @db['db_version'].to_i || 0
    end
  end
end
